lib/libc/stdlib/malloc/mrs/mrs.c

```diff
@@ -981,11 +992,46 @@ quarantine_flush(struct mrs_quarantine *quarantine)
 	struct mrs_descriptor_slab *prev = NULL;
 
 	MRS_UTRACE(UTRACE_MRS_QUARANTINE_FLUSH, NULL, 0, 0, NULL);
+
+	/*
+	 * If we're defering actual free calls until later, just add the
+	 * quaranting to the pending queue and let them be freed one at
+	 * a time by check_and_perform_flush().
+	 */
+	if (paced_free) {
+		struct mrs_quarantine *tmp;
+
+		mrs_lock(&app_quarantine_lock);
+		tmp = TAILQ_FIRST(&app_quarantine_free_list);
+		if (tmp == NULL) {
+			/*
+			 * XXX: It's possible there aren't any free
+			 * quarantines.  If that happens, punt and
+			 * flush this whole batch.  This will likely
+			 * blow the malloc caches, but prevents unbounded
+			 * growth.
+			 *
+			 * XXX: it might be better to flush the old one
+			 * instead, but that logic is more complicated.
+			 */
+			mrs_unlock(&app_quarantine_lock);
+			goto just_flush;
+		}
+		TAILQ_REMOVE(&app_quarantine_free_list, tmp, next);
+		quarantine_move(tmp, quarantine);
+		TAILQ_INSERT_TAIL(&app_quarantine_pending_list, tmp, next);
+		mrs_unlock(&app_quarantine_lock);
+
+		return;
+	}
+
+just_flush:
 	for (struct mrs_descriptor_slab *iter = quarantine->list; iter != NULL;
 	     iter = iter->next) {
+		void *ptr;
 		assert(iter->next_descriptor == 0);
-		while(iter->next_descriptor < iter->num_descriptors) {
-			slab_free_next_descriptor(iter);
+		while((ptr = slab_pop_descriptor(iter)) != NULL) {
+			descriptor_free(ptr);
 		}
 		assert(iter->next_descriptor == iter->num_descriptors);
```

Comment:
Shouldn't this get hit because we increment next_descriptor on the iteration where slab_pop_descriptor returns NULL

Reply:
I've updated slab_dequeue_descriptor to return NULL without incrementing.